networkit/cpp/embedding/BiasedRandomWalk.cpp

```diff
@@ -0,0 +1,154 @@
+        double pSum = 0;
+        std::vector<float> pTable; // Probability distribution table
+        // for each v-neighbor
+        graph->forNeighborsOf(v, [&](node x, edgeweight weight) {
+            if (x == t) {
+                pTable.push_back(weight / paramP);
+                pSum += weight / paramP;
+            } else if (tNbrs.count(x)) {
+                pTable.push_back(weight);
+                pSum += weight;
+            } else {
+                pTable.push_back(weight / paramQ);
+                pSum += weight / paramQ;
+            }
+        });
+        // Normalizing table
+        float pfSum = (float)pSum;
+        std::for_each(pTable.begin(), pTable.end(), [pfSum](float &p) { p /= pfSum; });
+
+        graphData->data[v][t].unigram(pTable);
+    }
+}
+
+// Preprocess transition probabilities for each path t->v->x
+void BiasedRandomWalk::preprocessTransitionProbs(double paramP, double paramQ) {
+    graph->balancedParallelForNodes([&](node t) { preprocessNode(t, paramP, paramQ); });
+}
+
+// Simulates a random walk
+Walk BiasedRandomWalk::oneWalk(node start, count walkLen) {
+    Walk walk(walkLen);
+    count nr = 0;
+    walk[nr++] = start;
+    node src = start;
+
+    if (walkLen == 1) {
+        return walk;
+    }
+    if (graph->degreeOut(start) == 0) {
+        walk.resize(1); // shorten walk to 1
+        return walk;
+    }
+    auto nn = graph->degree(start);
+
+    node randomNeighbor = index2node[start][nn * Aux::Random::real()];
+
+    walk[nr++] = randomNeighbor;
+    node dst = randomNeighbor;
+
+    while (nr < walkLen) {
+        if (graph->degreeOut(dst) == 0) {
+            walk.resize(nr); // shorten walk to nr
+            return walk;
+        }
+        NeighborMap &map = graphData->data[dst];
+        AliasSampler &as = map[src];
+        node next = index2node[dst][as.sample()];
+        walk[nr++] = next;
+        src = dst;
+        dst = next;
+    }
+    return walk;
+}
+
+struct WalkData {
+    // assuming contiguous node numbers 0..n-1, they serve as indices
+    AllWalks data;
+    count walkLength;
+    count walksPerNode;
+
+    WalkData(count nn, count wl, count wpn)
+        : data{nn * wpn, Walk(wl)}, walkLength(wl), walksPerNode(wpn) {}
+};
+
+/// Simulates walks from every node and writes it into Walks vector
+AllWalks BiasedRandomWalk::doWalks(count walkLen, count walksPerNode) {
+    auto nn = graph->numberOfNodes();
+    count walksDone = 0;
+
+    WalkData walkData(nn, walkLen, walksPerNode);
+    std::vector<node> shuffled(graph->nodeRange().begin(), graph->nodeRange().end());
+
+    for (index c = 0; c < walksPerNode; ++c) {
+        std::shuffle(shuffled.begin(), shuffled.end(), Aux::Random::getURNG());
+        graph->balancedParallelForNodes([&](node i) {
+            auto v = shuffled[i];
+            Walk thisWalk = oneWalk(v, walkLen);
+            walkData.data[c * nn + i] = std::move(thisWalk);
+            ++walksDone;
```

Comment:
This must be an atomic increment or some increments to `walksDone` will be lost. However, I also cannot see where `walksDone` is used - can we simply remove this?

Reply:
You are right, it's a relict of a former progress indicator. --> removed